File: modules/stego-core/src/main/java/com/shterneregen/securelan/stego/service/SteganographyService.java
package com.shterneregen.securelan.stego.service;

import com.shterneregen.securelan.stego.model.BmpCapacity;
import com.shterneregen.securelan.stego.model.ExtractedStegoPayload;
import com.shterneregen.securelan.stego.model.StegoContentType;

public interface SteganographyService {
    BmpCapacity inspect(byte[] bmpBytes);

    byte[] hide(byte[] bmpBytes, byte[] payload, StegoContentType contentType);

    ExtractedStegoPayload extract(byte[] bmpBytes);

    default byte[] hidePayload(byte[] bmpBytes, byte[] payload) {
        return hide(bmpBytes, payload, StegoContentType.BINARY);
    }

    byte[] extractPayload(byte[] bmpBytes);

    byte[] hideText(byte[] bmpBytes, String message);

    String extractText(byte[] bmpBytes);

    byte[] hideEncryptedPayload(byte[] bmpBytes, byte[] payload, char[] password);

    byte[] extractEncryptedPayload(byte[] bmpBytes, char[] password);

    byte[] hideEncryptedText(byte[] bmpBytes, String message, char[] password);

    String extractEncryptedText(byte[] bmpBytes, char[] password);
}

File: modules/stego-core/src/main/kotlin/com/shterneregen/securelan/stego/service/SteganographyService.kt
package com.shterneregen.securelan.stego.service

import com.shterneregen.securelan.stego.model.BmpCapacity
import com.shterneregen.securelan.stego.model.ExtractedStegoPayload
import com.shterneregen.securelan.stego.model.StegoContentType

interface SteganographyService {
    fun inspect(bmpBytes: ByteArray): BmpCapacity

    fun hide(bmpBytes: ByteArray, payload: ByteArray, contentType: StegoContentType): ByteArray

    fun extract(bmpBytes: ByteArray): ExtractedStegoPayload

    fun hidePayload(bmpBytes: ByteArray, payload: ByteArray): ByteArray = hide(bmpBytes, payload, StegoContentType.BINARY)

    fun extractPayload(bmpBytes: ByteArray): ByteArray

    fun hideText(bmpBytes: ByteArray, message: String): ByteArray

    fun extractText(bmpBytes: ByteArray): String

    fun hideEncryptedPayload(bmpBytes: ByteArray, payload: ByteArray, password: CharArray): ByteArray

    fun extractEncryptedPayload(bmpBytes: ByteArray, password: CharArray): ByteArray

    fun hideEncryptedText(bmpBytes: ByteArray, message: String, password: CharArray): ByteArray

    fun extractEncryptedText(bmpBytes: ByteArray, password: CharArray): String
}
